tokio/src/runtime/basic_scheduler.rs

```diff
@@ -117,13 +136,75 @@ where
         self.spawner.spawn(future)
     }
 
-    pub(crate) fn block_on<F>(&mut self, future: F) -> F::Output
-    where
-        F: Future,
-    {
+    pub(crate) fn block_on<F: Future>(&self, future: F) -> F::Output {
+        // If we can steal the dedicated parker then lets block_on that
+        // Otherwise, we'll block_on the future and attempt to steal the
+        // parker later, if we can.
+        if let Some(mut inner) = self.take_inner() {
+            inner.block_on(future)
+        } else {
+            let enter = crate::runtime::enter(false);
+
+            let mut park = CachedParkThread::new();
+            let waker = park.unpark().into_waker();
+            let mut cx = std::task::Context::from_waker(&waker);
+
+            pin!(future);
+
+            let notifier = {
+                let lock = self.inner.lock().unwrap();
+                lock.notify.clone()
+            };
+
+            let mut notified = Box::pin(notifier.notified());
+
+            loop {
+                if let Ready(_) = notified.as_mut().poll(&mut cx) {
+                    // Check if we can steal the dedicated parker P.
+                    //
+                    // TODO: Consider using an atomic load here intead of locking
+                    // the mutex.
+                    if let Some(mut inner) = self.take_inner() {
+                        // We will enter again on in the inner implementation below
+                        drop(enter);
+                        return inner.block_on(future);
+                    } else {
+                        // Since the notify future polled to ready, it will panic if polled again
+                        // beyond ready. To avoid this, lets create a new future. This allocation is
+                        // unfortunate but should be extremely rare.
+                        notified = Box::pin(notifier.notified());
+                    }
+                }
+
+                if let Ready(v) = crate::coop::budget(|| future.as_mut().poll(&mut cx)) {
+                    return v;
+                }
+
+                // Park this thread, waiting for some external wakeup: either
+                // from the future we are currently polling or a wakeup from the
+                // block_on that contains the parker, notifying us to steal the parker.
+                park.park().expect("failed to park");
+            }
+        }
+    }
```

Comment:
I believe this can be simplified as something like:

```rust
pub fn block_on(&self, future: F) -> F::Output {
    tokio::pin!(future);

    loop {
        // Try to take the scheduler
        if let Some(mut inner) = self.take_inner() {
            return inner.block_on(&mut future);
        }

        let enter = crate::runtime::enter(false);
        let out = enter.block_on(async {
            select! {
                out = &mut future => Some(out),
                _ => self.notify.notified() => None,
            }
        });

        if let Some(out) = out {
            return out;
        }
    }
}
```

Reply:
Unable to use macros within the tokio crate because https://github.com/rust-lang/rust/pull/52234

Reply:
Does this work?

```rust
pub fn block_on(&self, future: F) -> F::Output {
    tokio::pin!(future);

    loop {
        // Try to take the scheduler
        if let Some(mut inner) = self.take_inner() {
            return inner.block_on(&mut future);
        }

        let notified = self.notify.notified();
        tokio::pin!(notified);

        let enter = crate::runtime::enter(false);
        let out = enter.block_on(poll_fn(|cx| {
            if notified.poll(cx).is_ready() {
                return Ready(None);
            }

            if let Ready(out) = future.poll(cx) {
                return Ready(Some(out));
            }

            Pending
        }));

        if let Some(out) = out {
            return out;
        }
    }
}
```